components/eventing-controller/pkg/object/equality.go

```diff
@@ -48,3 +58,253 @@ func apiRuleEqual(a1, a2 *apigatewayv1alpha1.APIRule) bool {
 
 	return true
 }
+
+// eventingBackendEqual asserts the equality of two EventingBackend objects.
+func eventingBackendEqual(b1, b2 *eventingv1alpha1.EventingBackend) bool {
+	if b1 == nil || b2 == nil {
+		return false
+	}
+
+	if b1 == b2 {
+		return true
+	}
+
+	if !reflect.DeepEqual(b1.Labels, b2.Labels) {
+		return false
+	}
+
+	if !reflect.DeepEqual(b1.Spec, b2.Spec) {
+		return false
+	}
+
+	return true
+}
+
+// publisherProxyDeploymentEqual asserts the equality of two Deployment objects for event publisher proxy deployments.
+func publisherProxyDeploymentEqual(d1, d2 *appsv1.Deployment) bool {
+	if d1 == nil || d2 == nil {
+		return false
+	}
+	if d1 == d2 {
+		return true
+	}
+
+	if !reflect.DeepEqual(d1.Labels, d2.Labels) {
+		return false
+	}
+
+	cst1 := d1.Spec.Template
+	cst2 := d2.Spec.Template
+
+	if !reflect.DeepEqual(cst1.Annotations, cst2.Annotations) {
+
+		return false
+	}
+
+	if !reflect.DeepEqual(cst1.Labels, cst2.Labels) {
+		return false
+	}
+
+	ps1 := &cst1.Spec
+	ps2 := &cst2.Spec
+	if !podSpecEqual(ps1, ps2) {
+		return false
+	}
+
+	return true
+}
+
+// podSpecEqual asserts the equality of two PodSpec objects.
+func podSpecEqual(ps1, ps2 *corev1.PodSpec) bool {
+	if ps1 == nil || ps2 == nil {
+		return false
+	}
+	if ps1 == ps2 {
+		return true
+	}
+
+	cs1, cs2 := ps1.Containers, ps2.Containers
+	if len(cs1) != len(cs2) {
+		return false
+	}
+	for i := range cs1 {
+		if !containerEqual(&cs1[i], &cs2[i]) {
+			return false
+		}
+	}
+
+	if ps1.ServiceAccountName != ps2.ServiceAccountName {
+		return false
+	}
+
+	return true
+}
+
+// containerEqual asserts the equality of two Container objects.
+func containerEqual(c1, c2 *corev1.Container) bool {
+	if c1 == nil || c2 == nil {
+		return false
+	}
+	if c1.Image != c2.Image {
+		return false
+	}
+
+	ps1, ps2 := c1.Ports, c2.Ports
+	if len(ps1) != len(ps2) {
+		return false
+	}
+	for i := range ps1 {
+		p1, p2 := &ps1[i], &ps2[i]
+
+		if p1.Name != p2.Name ||
+			p1.ContainerPort != p2.ContainerPort ||
+			realProto(p1.Protocol) != realProto(p2.Protocol) {
+
+			return false
+		}
+	}
+
+	if !envEqual(c1.Env, c2.Env) {
+		return false
+	}
+
+	if !probeEqual(c1.ReadinessProbe, c2.ReadinessProbe) {
+		return false
+	}
+
+	return true
+}
+
+func envEqual(e1, e2 []corev1.EnvVar) bool {
+	if len(e1) != len(e2) {
+		return false
+	}
+EV1:
+	for _, ev1 := range e1 {
+		for _, ev2 := range e2 {
+			if reflect.DeepEqual(ev1, ev2) {
+				continue EV1
```

Comment:
(nit) we can achieve the same with `break`.

Reply:
> remove EV1: at line 182
> use break instead of continue EV1 at line 186
> remove return false at line 189

Not really. It'd never return `false` in case of mismatch of env vars.

Reply:
You're right, I forgot to mention something like keeping a match count:

```golang
func envEqual(e1, e2 []corev1.EnvVar) bool {
	if len(e1) != len(e2) {
		return false
	}

	matchCount := 0
	for _, ev1 := range e1 {
		for _, ev2 := range e2 {
			if reflect.DeepEqual(ev1, ev2) {
				matchCount++
				break
			}
		}
	}
	return matchCount == len(e1)
}
```